posts/2018-04-13-rethinking-monaderror.md

```diff
@@ -0,0 +1,463 @@
+  def raiseError[A](e: E): F[A]
+}
+```
+
+This definition looks identical to the one defined one `MonadError` so let's move on to error-handling.
+For handled errors, we want to return a value inside `G`, so our `handleErrorWith` function should indeed return a `G[A]`:
+
+```scala
+trait MonadBlunder[F[_], G[_], E] {
+  ...
+
+  def handleErrorWith[A](fa: F[A])(f: E => F[A]): G[A]
+}
+```
+
+Looks good so far, right? 
+Well, we still have the problem that `f` might return an erronous value, so if we want to guarantee that the result won't have any errors, we'll have to change that to `G[A]` as well:
+
+```scala
+trait MonadBlunder[F[_], G[_], E] {
+  ...
+
+  def handleErrorWith[A](fa: F[A])(f: E => G[A]): G[A]
+}
+```
+
+And now we're off to a pretty good start, we fixed one short coming of `MonadError` with this approach.
+
+Another approach, maybe more obvious to some, might be to require the type constructor to take two arguments, one for the value and one for the error type.
+Let's see if we can define `raiseError` on top of it:
+
+```scala
+trait MonadBlunder[F[_, _]] {
+  def raiseError[E, A](e: E): F[E, A]
+
+  ...
+}
+```
+
+This looks pretty similar to what we already have, though now we have the guarantee that our type doesn't actually "hide" the error-type somewhere.
+Next up is `handleErrorWith`. Ideally after we handled the error we should again get back a type that signals that it doesn't have any errors. 
+We can do exactly that by choosing an unhabited type like `Nothing` as our error-type:
+
+
+```scala
+trait MonadBlunder[F[_, _]] {
+  ...
+
+  def handleErrorWith[E, A](fa: F[E, A])(f: E => F[Nothing, A]): F[Nothing, A]
+}
+```
+
+And this approach works as well, however now we've forced the two type parameter shape onto implementors. This `MonadBlunder` has the following kind `(* -> * -> *) -> * -> *`.
+This means we can very easily define instances for types with two type parameters like `Either`.
+However, one issue might be that it's much easier to fit a type with two type parameters onto a type class that expects a single type constructor `(* -> *)` than to do it the other way around.
+
+For example try to implement the above `MonadBlunder[F[_, _]]` for the standard `cats.effect.IO`.
+It's not going to be simple, whereas with the first encoding we can easily encode both `Either` and `IO`. For this reason, I will continue this article with the first encoding using the two different type constructors.
+
+Next we're going to look at laws we can define to make sense of the behaviour we want.
+The first two laws should be fairly obvious. 
+If we `flatMap` over a value created by `raiseError` it shouldn't propogate:
+
+```scala
+def raiseErrorStops(e: E, a: A): Boolean =
+  F.raiseError[A](e).flatMap(_ => a.pure[F]) === F.raiseError[A](e)
+```
+
+In the same way, we want values of `G` to never stop propagating, so `flatMap` should always work:
+
+```scala
+def flatMapAlwaysWorks(ga: G[A], a: A): Boolean =
```

Comment:
I don't understand this; this seems to be the exact opposite of the monad law for `flatMap` and `pure`. Except for the extra `a`, which actually makes this law guaranteed by parametricity as well for any monad other than the `Void[A] = Nothing` monad.

Reply:
The last bit there I'm not sure about. This law just confuses me.

Reply:
Yeah, as I said above, it's not really satisfying right now. I just want a law that prevents an error-ful type in `G`. Without this law I can define instances like `MonadBlunder[Task, Task, Throwable]`, which sort of defeats the point I guess.

Reply:
To express what you want, I think you want a typeclass method that enables _injecting_ `G` into `F` (`def inject[A](g: G[A]): F[A]`) — you seem to imply that's possible anyway. Then you can try to state that handling errors in `inject(g)` never has anything to handle — one way to write that might be `handleErrorWith(inject(ga))(f) = ga`. @LukaJCB @edmundnoble WDYT?

Reply:
That's excellent, thank you!